drivers/drive-ks.cpp

```diff
@@ -0,0 +1,1412 @@
+                       int ntime_,
+                       int target_nt,
+                       int cfactor_,
+                       int cfactor0_,
+                       int DeltaLevel_,
+                       int DeltaRank_,
+                       bool cglv_,
+                       bool useTheta_,
+                       bool doRefine_,
+                       bool getInit_,
+                       bool output_,
+                       int newton_iters_,
+                       int max_levels_,
+                       int nx,
+                       double length,
+                       int order,
+                       const VEC &initial_data_) : BraidApp(comm_t_, tstart_, tstop_, ntime_)
+{
+   rank = rank_;
+   cfactor = cfactor_;
+   cfactor0 = cfactor0_;
+   newton_iters = newton_iters_;
+   DeltaRank = DeltaRank_;
+   DeltaLevel = DeltaLevel_;
+   delayDelta = 0;
+   cglv = cglv_;
+   useTheta = useTheta_;
+   doRefine = doRefine_;
+   getInit = getInit_;
+   output = output_;
+   max_levels = max_levels_;
+   est_norm = -1.;
+   int worldsize, time_chunk;
+   MPI_Comm_size(comm_t, &worldsize);
+   time_chunk = target_nt / worldsize + 1;
+   err_ests.assign(time_chunk, 0.);
+
+   switch (order)
+   {
+   case 2:
+      stages = 2;
+      break;
+
+      // todo: implement 4th order method
+      // case 4:
+      //    stages = 3;
+      //    break;
+
+   default:
+      throw std::invalid_argument("Order not implemented");
+      break;
+   }
+
+   // fourth order
+   Stencil d1 = Stencil({ 1. / 12, -2. / 3, 0., 2. / 3, -1. / 12 });
+   Stencil d2 = Stencil({ -1. / 12, 4. / 3, -5. / 2, 4. / 3, -1. / 12 });
+   Stencil d4 = Stencil({ -1. / 6, 2., -13. / 2, 28. / 3, -13. / 2, 2., -1. / 6 });
+   disc = KSDiscretization(nx, length, d1, d2, d4);
+   initial_data = initial_data_;
+
+   // compute theta based on effective grid level
+   thetas.assign(max_levels, 1.);
+   if (useTheta)
+   {
+      double total_cf;
+      for (int level = 1; level < max_levels; level++)
+      {
+         total_cf = cfactor0 * intpow(cfactor, level - 1);
+
+         if (order == 2)
+         {
+            thetas[level] = (total_cf - sqrt(3 * total_cf * total_cf + 6) / 3) / total_cf; // theta2
+         }
+
+         if (order == 4)
+         {
+            double cf_4 = std::pow(total_cf, 4);
+            if (cf_4 == 0.) // overflow
+            {
+               thetas[level] = 1 - std::sqrt(10) / 5.;
+            }
+            else
+            {
+               thetas[level] = 1. - std::sqrt(10 * cf_4 + 15) / (5 * total_cf * total_cf);
+            }
+         }
+      }
+   }
+}
+
+// Helper function to check if current point is a C point for this level
+bool
+MyBraidApp::IsCPoint(int i, int level)
+{
+   if (level == 0)
+   {
+      return ((i % cfactor0) == 0);
+   }
+   return ((i % cfactor) == 0);
+}
+
+// Helper function to compute theta for a given level
+double
+MyBraidApp::getTheta(int level)
+{
+   return thetas[level];
+}
+
+void
```

Comment:
The following 3 or 4 theta-themed functions don't have a comment describing what they do.  Could a little more commentary be added?